### ISSUE TITLE
Structure pruning is broken for models with non-serial connections

### ISSUE THREAD

**Issue by nzmora** [NervanaSystems/distiller#85]
Structure pruning is broken for models with non-serial connections.
Models such as Alexnet and VGG are have serial data-dependencies (connections) and are fine.
More complex models, with parallel-data dependencies (paths), such as ResNets (skip connections) and GoogLeNet (Inception layers) might fail when pruning filters or channels.

This is because a module, such as ```torch.nn.modules.batchnorm.BatchNorm2d``` layers, might depend on multiple inputs.  This is not always a problem.  For example, if the dependent module has type ```torch.nn.Conv2d``` and we are pruning weight filters.
But if the dependent module has type ```torch.nn.modules.batchnorm.BatchNorm2d```, and we are pruning weight filters, then it is possible that each of the inputs selects different activation channels to prune.  In such a case, how should we prune the BatchNorm's scale and shift tensors (```.weight``` and ```.bias```)?

To solve this we need to define one of the modules as the *leader* which determines what activation channels to prune; and define the rest of the modules in the dependency sub-graph as *followers*.  *Followers* do not choose which activation channels to prune, so their sparsity masks is determined by the choice of the *leader*.  
Because the sparsity maps of different *follower* modules may have different shapes, the *leader* defines a **binary map** which is a binary vector of active (1) and pruned (0) channels.  Each "follower" expands this single **binary map** to create its own private **pruning mask**.

This requires changing the way we express filter/channel pruning in YAML, and how we create **pruning masks**.

I'm trying to make this fix available soon. 
This is related to issues #79 and #73.

**Comment by nzmora** [comment]
@cattpku @vinutah @chenys1995 @promach 
I merged to 'master' a PR with the fix for issue #85 which should also resolve #79 #86 #73 #36.

There are YAML API for structured-pruning changes with examples.  I also added some explanations about the API here:
https://github.com/NervanaSystems/distiller/wiki/Pruning-Filters-&-Channels

**Comment by chenys1995** [comment]
I am confuse about the "concat1" in pruning_structs_ex4.png
It seems like a element-wise add
In the case of the channel-wise concatenation
 (1) conv1 -> conv3
 (2) conv2 -> conv3(with offset index)
can be done without any confusion(Just my imagine)

**Comment by nzmora** [comment]
Hi @chenys1995,

Regarding your [question](https://github.com/NervanaSystems/distiller/issues/85#issuecomment-443478843): I read through the example I gave and I can see that in the diagram I showed an element-wise summation, and in the text I referred to concatenation and that this is confusing and wrong.  So this is now fixed.

BTW, if instead of performing element-wise addition we used concatenation, then we would have an even more complex dependency: from the weights of ```conv3``` we would need to remove the channels corresponding to the filters pruned by **both** ```conv1``` **and** ```conv2```.  This is not supported yet.

Thanks,
Neta

**Comment by promach** [comment]
> Hi @promach,
> 
> Regarding the [problem you see ](https://github.com/NervanaSystems/distiller/issues/85#issuecomment-443472825): it appears to be caused by PyTorch 0.4.1 which is not supported by Distiller yet.
> 
> Cheers
> Neta

@nzmora I am not being pushy and I am also not contributing towards distiller development. I am just a normal user. May I know why distiller still not supporting latest pytorch ? Most systems are now using pytorch version 1.0 ?

Please do not take my question as an attack. I just wish to use distiller on my system.

**Comment by nzmora** [comment]
Hi @promach,

This is a fair question.  We've been planning support for PyTorch 1.0 for some time, but it wasn't high on our priority list.  Anyway, I hope to be able to release PyTorch 1.0 support in a few weeks.

I'm sorry you are having issues installing PyTorch 0.4.  This version is still supported and using ```requirements.txt``` should work for you.  
The easiest way to list the available versions of a package is to try installing an illegal version.  The error message lists the available versions and you can see 0.4 is supported:
```
$ pip3 install torch==700
Collecting torch==700
  Could not find a version that satisfies the requirement torch==700 (from versions: 0.1.2, 0.1.2.post1, 0.3.1, 0.4.0, 0.4.1, 1.0.0)
No matching distribution found for torch==700
```

You can also see it listed in PyPi: https://pypi.org/project/torch/0.4.0/

Please let me know if you figure out why you can't install pytorch 0.4 on your system - that should not be the case.
Cheers,
Neta

**Comment by promach** [comment]
Thanks @nzmora for trying to support pytorch version 1.0

> [phung@archlinux distiller]$ source env/bin/activate
> (env) [phung@archlinux distiller]$ pip install torch==0.4.0
> Collecting torch==0.4.0
>   Could not find a version that satisfies the requirement torch==0.4.0 (from versions: 0.1.2, 0.1.2.post1, 0.4.1, 0.4.1.post2, 1.0.0)
> No matching distribution found for torch==0.4.0
> (env) [phung@archlinux distiller]$ pip install -r requirements.txt
> Collecting torch==0.4.0 (from -r requirements.txt (line 1))
>   Could not find a version that satisfies the requirement torch==0.4.0 (from -r requirements.txt (line 1)) (from versions: 0.1.2, 0.1.2.post1, 0.4.1, 0.4.1.post2, 1.0.0)
> No matching distribution found for torch==0.4.0 (from -r requirements.txt (line 1))
> (env) [phung@archlinux distiller]$

**Comment by promach** [comment]
See the log below for whl package incompatibility with my system-wide python

Take your time slowly to do the development stuff for pytorch version 1.0 support. I am only a casual user, planning to use distiller once it is ready for pytorch 1.0 in the near future.

> [phung@archlinux distiller]$ source env/bin/activate
> (env) [phung@archlinux distiller]$ ls env/
> bin  include  lib  pip-selfcheck.json  torch-0.4.0-cp36-cp36m-manylinux1_x86_64.whl
> (env) [phung@archlinux distiller]$ pip install env/torch-0.4.0-cp36-cp36m-manylinux1_x86_64.whl
> torch-0.4.0-cp36-cp36m-manylinux1_x86_64.whl is not a supported wheel on this platform.
> (env) [phung@archlinux distiller]$ python
> Python 3.7.2 (default, Dec 29 2018, 21:15:15) 
> [GCC 8.2.1 20181127] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> quit
> Use quit() or Ctrl-D (i.e. EOF) to exit
> >>> 
> (env) [phung@archlinux distiller]$ python

> python                                        python2-config                                python3.7-config                              python-argcomplete-check-easy-install-script
> python2                                       python2-pyuic4                                python3.7m                                    python-argcomplete-tcsh
> python2.7                                     python3                                       python3.7m-config                             python-config
> python2.7-config                              python3.7                                     python3-config                                
> (env) [phung@archlinux distiller]$ python
